### PR TITLE
Stop running Spark as a deamon when client mode is on.

### PR DESCRIPTION
**Pull request by cosmincatalin**
Allow using the image solely for submitting applications to other Spark clusters.

Using `--deploy-mode cluster` along with `docker-compose` will probably result in application failures, since the master needs to be up. The default deploy mode, `client` is best to be used for this particular purpose.

### CODE REVIEW

**Comment by andresbono** [comment]
Hi @cosmincatalin, thanks for opening this PR. We will review it and we will get back to you.

File: docker-compose.yml [hunk 1]
```diff
@@ -33,3 +33,19 @@ services:
...
+    command:
+      - spark-submit
+      - --deploy-mode
+      - client
+      - --class
+      - org.apache.spark.examples.SparkPi
+      - --master
+      - spark://spark:7077
+      - https://repo1.maven.org/maven2/org/apache/spark/spark-examples_2.10/1.1.1/spark-examples_2.10-1.1.1.jar
+      - "1000"
```
Comment (andresbono) [inline docker-compose.yml:51]: I would keep the snippet of the README, but I wouldn't include the client in the main docker-compose of the repository. This is using an example application that wouldn't be useful for a general purpose.

File: 3/debian-10/rootfs/opt/bitnami/scripts/libspark.sh [hunk 2]
```diff
@@ -78,10 +78,10 @@ spark_validate() {
 
     # Validate spark mode
     case "$SPARK_MODE" in
-        master|worker)
+        master|worker|client)
         ;;
         *)
-            print_validation_error "Invalid mode $SPARK_MODE. Supported types are 'master/worker'"
+            print_validation_error "Invalid mode $SPARK_MODE. Supported types are 'master/worker/client'"
```
Comment (andresbono) [inline 3/debian-10/rootfs/opt/bitnami/scripts/libspark.sh:84]: If this is not being used, I would remove it.

File: README.md [hunk 3]
```diff
@@ -94,7 +94,7 @@ $ docker run -d --name spark \
 
 Available variables:
 
-* SPARK_MODE: Cluster mode starting Spark. Valid values: *master*, *worker*. Default: **master**
+* SPARK_MODE: Cluster mode starting Spark. Valid values: *master*, *worker*, *client*. Default: **master**
```
Comment (andresbono) [inline README.md:97]: Same comment than above. If it is not being used, I would revert this change.

File: README.md [hunk 4]
```diff
@@ -210,6 +210,27 @@ $ pyspark
 '2.7.4'
 ```
 
+### Submitting applications
+
+When using the *client* mode, Spark will not run as a daemon. You can then use `spark-submit` to run jobs on a cluster.
+
+```yaml
+environment:
+  - SPARK_MODE=client
```
Comment (andresbono) [inline README.md:219]: It seems it is not needed in the end.

**Review by andresbono — COMMENTED** [review]
Thank you very much for your contribution! Please check my comments.

File: 3/debian-10/rootfs/opt/bitnami/scripts/spark/run.sh [hunk 5]
```diff
@@ -19,14 +19,19 @@ if [ "$SPARK_MODE" == "master" ]; then
...
-    exec "$EXEC" "${ARGS[@]-}"
-fi
+
+if [ "$SPARK_MODE" != "client" ]; then
+    if am_i_root; then
+        exec gosu "$SPARK_DAEMON_USER" "$EXEC" "${ARGS[@]-}"
+    else
+        exec "$EXEC" "${ARGS[@]-}"
+    fi
+fi
```
[the remark this answers is not in the document]
Comment (cosmincatalin) [reply 3/debian-10/rootfs/opt/bitnami/scripts/spark/run.sh:37]: Right, then I guess the PR looses its value then. Indeed you can use it directly.

**Comment by andresbono** [comment]
Hi @cosmincatalin, we still see value in the README.md changes. It would be nice to merge the `Submitting applications` section. Just in case you want to reopen and modify the PR.

**Comment by cosmincatalin** [comment]
Sure let me try to make a PR with just that

**Comment by andresbono** [comment]
Great, thanks.
